crates/compose/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+        &self,
+        source: &locked::LockedComponentSource,
+    ) -> anyhow::Result<Vec<u8>>;
+}
+
+struct Composer<'a, L> {
+    graph: CompositionGraph,
+    loader: &'a L,
+}
+
+impl<'a, L: ComponentLoader> Composer<'a, L> {
+    async fn compose(mut self, component: &LockedComponent) -> anyhow::Result<Vec<u8>> {
+        let source = self.loader.load_component(&component.source).await?;
+
+        if component.dependencies.is_empty() {
+            return Ok(source);
+        }
+
+        let package = Package::from_bytes(&component.id, None, source, self.graph.types_mut())?;
+
+        let world_id = package.ty();
+        let package_id = self.graph.register_package(package)?;
+        let instantiation_id = self.graph.instantiate(package_id);
+
+        let prepared = self.prepare_dependencies(world_id, component).await?;
+
+        let arguments = self
+            .build_instantiation_arguments(world_id, prepared)
+            .await?;
+
+        for (argument_name, argument) in arguments {
+            self.graph
+                .set_instantiation_argument(instantiation_id, &argument_name, argument)?;
+        }
+
+        // Export all exports from the root component
+        for export_name in self.graph.types()[world_id]
+            .exports
+            .keys()
+            .cloned()
+            .collect::<Vec<_>>()
+        {
+            let export_id = self
+                .graph
+                .alias_instance_export(instantiation_id, &export_name)?;
+
+            self.graph.export(export_id, &export_name)?;
+        }
+
+        self.graph
+            .encode(Default::default())
+            .with_context(|| "failed to encode composition graph")
+            .map(Into::into)
+    }
+
+    fn new(loader: &'a L) -> Self {
+        Self {
+            graph: CompositionGraph::new(),
+            loader,
+        }
+    }
+
+    async fn prepare_dependencies(
+        &mut self,
+        world_id: WorldId,
+        component: &LockedComponent,
+    ) -> anyhow::Result<IndexMap<String, DependencyInfo>> {
+        let mut expanded = IndexMap::new();
+
+        let mut imports = self.graph.types()[world_id].imports.clone();
+
+        // TODO: collect duplicates and return
+
+        for (dependency_name, dependency) in component.dependencies.iter() {
+            let import_keys = imports.keys().cloned().collect::<Vec<_>>();
+
+            let mut matched = Vec::new();
+
+            for import_name in import_keys.iter() {
+                if dependency_name.matches_import(import_name) {
+                    match imports.swap_remove(import_name) {
+                        Some(_item) => {
+                            matched.push(import_name.clone());
+                            break;
+                        }
+                        None => {
+                            anyhow::bail!(
+                                "import '{import_name}' of component '{component_id}' is already used by another dependency",
```

Comment:
I think we'll want to workshop these errors a bit as I think they come across a bit more technical than we may want.

Reply:
While pairing on this @fibonacci1729 said to me "I'll let people with strong opinions about error messages comment on the PR".

Reply:
It might be nice to frame these as "conflicts" between pairs of dependencies. We also discussed collecting all conflicts before bailing to give more comprehensive conflict reports.

Reply:
Yeah this is definitely an area where i welcome feedback from our wordsmiths.

Reply:
+1 on the user facing error here being more descriptive.

Reply:
The latest commit switches to using `thiserror` to return a typed error. I'm sure the descriptions could be improved so let me know if you have ideas.